internal/wasm/store.go

```diff
@@ -451,13 +452,17 @@ func (s *Store) instantiate(
 	}
 
 	// Compile the default context for calls to this module.
-	m.CallCtx = NewCallContext(ns, m, sys)
+	m.CallCtx = NewCallContext(ns, m, sysCtx)
 
 	// Execute the start function.
 	if module.StartSection != nil {
 		funcIdx := *module.StartSection
 		f := m.Functions[funcIdx]
-		if _, err = f.Module.Engine.Call(ctx, m.CallCtx, f); err != nil {
+		_, err = f.Module.Engine.Call(ctx, m.CallCtx, f)
+
+		if exitErr, ok := err.(*sys.ExitError); ok { // Don't wrap an exit error!
```

Comment:
Just wondering, would it make sense to not propagate code == 0 error here? Looking at the doc for `ExitError` I do see it mention the code == 0 case possibly being ok

https://github.com/tetratelabs/wazero/blob/main/sys/error.go#L15

But I never noticed it 😅 With Go muscle memory I generally assume that `err != nil` is a failure of some sort. `io.EOF` is similar so there's precedent, but same muscle memory always trips me up on that too.

Reply:
I was wondering the same thing this morning. Here's the only rationale I can think of that works.

* For a no return function, there's no issue except if you want to know if someone called exit or not. For example, it isn't always expected to call exit.
* For a function with a return, there's an issue as now you don't have a return value that you need.

What we could do to split the difference is that for the "start functions" we can swallow exit code zero, but otherwise not. Ex if you call a function with 1 return and that eventually calls exit, don't interfere.

If someone really wants to see exit code zero they can tell us between now and 1.0 even after beta starts end of aug.

Reply:
exit code zero is uncommon even if emscripten will return it. I checked tinygo and zig compiled wasi-libc. Since it invalidates the module (causing future functions to be broke), I think the best thing is to document the rare case chance of an implicit exit zero and hope it stays rare!